src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -776,6 +777,132 @@ bool MetaspaceShared::try_link_class(InstanceKlass* ik, TRAPS) {
+      return;
+    }
+    if (file_name == nullptr) {
+      os::snprintf(filename, sizeof(filename), "java_pid%d_dynamic.jsa", os::current_process_id());
+      file = filename;
+    } else {
+      if (strstr(file_name, ".jsa") == nullptr) {
+        os::snprintf(filename, sizeof(filename), "%s.jsa", file_name);
+        file = filename;
+      }
+    }
+    cmd_dump_dynamic(output, file, THREAD);
+  }
+}
+
+class DumpClassListCLDClosure : public CLDClosure {
+  fileStream *_stream;
+public:
+  DumpClassListCLDClosure(fileStream* f) : CLDClosure() { _stream = f; }
+  ~DumpClassListCLDClosure() {
+    delete _stream; // The file need close since in child process it will be used.
+  }
+  void do_cld(ClassLoaderData* cld) {
+    for (Klass* klass = cld->klasses(); klass != NULL; klass = klass->next_link()) {
+      if (klass->is_instance_klass()) {
+        InstanceKlass* ik = InstanceKlass::cast(klass);
+        if (ik->is_shareable()) {
+          _stream->print_cr("%s", ik->name()->as_C_string());
+        }
+      }
+    }
+  }
+};
+
+//   To create a static dump, steps:
+//   1. output a classlist file
+//   2. fork a new process to dump the shared archive
+void MetaspaceShared::cmd_dump_static(outputStream* output, const char* file, TRAPS) {
+  const char* java_home = Arguments::get_java_home();
+  const char* file_separator = os::file_separator();
+  const char* app_class_path = Arguments::get_appclasspath();  // -cp ..
+  const char* java_command = Arguments::java_command();        // arguments to app
+  char exec_path[JVM_MAXPATHLEN]; // $JAVA_HOME/bin/java or %JAVA_HOME%\\jre\bin\\java ...
+  char classlist_name[JVM_MAXPATHLEN];
+
+  os::snprintf(classlist_name, sizeof(classlist_name), "%s.classlist", file);
+
+  ResourceMark rm;
+  fileStream* stream = new (ResourceObj::C_HEAP, mtInternal) fileStream(classlist_name, "w");
+  if (stream->is_open()) {
+    MutexLocker lock(ClassLoaderDataGraph_lock);
+    DumpClassListCLDClosure collect_classes(stream);
+    ClassLoaderDataGraph::loaded_cld_do(&collect_classes);
+  } else {
+    output->print_cr("Error to open %s for write!", classlist_name);
+    return;
+  }
+  os::snprintf(exec_path, sizeof(exec_path),
+               "%s%sbin%sjava -Xshare:dump -XX:SharedClassListFile=%s -XX:SharedArchiveFile=%s ",
+               java_home, file_separator, file_separator, classlist_name, file);
+  int num_vm_args = Arguments::num_jvm_args();
+  char** vm_args  = Arguments::jvm_args_array();
+  for (int i = 0; i < num_vm_args; i ++) {
+    append_strings(exec_path, sizeof(exec_path), vm_args[i]);
+  }
+  // Turn off RecordDynamicDumpInfo
+  if (RecordDynamicDumpInfo) {
+    append_strings(exec_path, sizeof(exec_path), "-XX:-RecordDynamicDumpInfo");
+  }
+  char* buff_start = exec_path + strlen(exec_path);
+  snprintf(buff_start, sizeof(exec_path), " -cp %s %s", app_class_path, java_command);
+  output->print_cr("%s", exec_path);
+  os::fork_and_exec(exec_path);
```

Comment:
Apart from what I wrote above, how do you handle fork/exec errors? There are a number of things that can go wrong here. At the very least I would scan the return code. And the return code of the child VM.